### PR TITLE
Upload flow refactor

### PR DESCRIPTION
**Pull request by amueller**
work in progress

### CODE REVIEW

**Comment by amueller** [comment]
There's currently a test failure because the order of the keys in ``{"step_name": "ohe", "key": "steps__ohe"}`` is changed. That should probably also be an ordered dict, not sure if I broke that here somewhere.

**Comment by amueller** [comment]
I'm gonna work on the scikit-learn conversion tomorrow, time for dinner now.

File: tests/flows/test_flow.py [hunk 1]
```diff
@@ -205,12 +184,10 @@ def side_effect(self):
             server_xml = server_xml.replace('  ', '').replace('\t', '').strip().replace('\n\n', '\n').replace('&quot;', '"')
             server_xml = re.sub(r'^$', '', server_xml)
 
-
         self.assertEqual(server_xml, local_xml)
 
         self.assertEqual(new_flow, flow)
         self.assertIsNot(new_flow, flow)
-        new_flow.model.fit(X, y)
```
Comment (mfeurer) [inline tests/flows/test_flow.py]: Why did you remove this fit?

Comment (mfeurer) [inline tests/flows/test_flow.py]: Why did you remove this fit?

**Comment by amueller** [comment]
I'll have to double check the tests. Sorry I'm slow...

File: openml/flows/flow.py [hunk 2]
```diff
@@ -154,38 +155,28 @@ def _to_dict(self):
...
-        if self.upload_date is not None:
-            flow_dict['oml:flow']['oml:upload_date'] = self.upload_date
-        if self.language is not None:
-            flow_dict['oml:flow']['oml:language'] = self.language
-        if self.dependencies is not None:
-            flow_dict['oml:flow']['oml:dependencies'] = self.dependencies
+        flow_container = OrderedDict()
+        flow_dict = OrderedDict([('@xmlns:oml', 'http://openml.org/openml')])
+        flow_container['oml:flow'] = flow_dict
+        _add_if_nonempty(flow_dict, 'oml:id', self.flow_id)
```
Comment (amueller) [inline openml/flows/flow.py]: because it's called ``flow_id`` not ``id``

File: openml/flows/flow.py [hunk 3]
```diff
@@ -154,38 +155,28 @@ def _to_dict(self):
...
+
+        for attribute in ["uploader", "name", "version", "external_version",
```
Comment (amueller) [inline openml/flows/flow.py]: done

File: tests/flows/test_flow.py (hunk 1 above)
Comment (amueller) [inline tests/flows/test_flow.py]: I moved it below the asserts, I guess because I felt that it was a better order to have all the string asserts in the same place? I don't have a strong opinion, though.

**Comment by amueller** [comment]
Hm this error looks odd again....

**Comment by mfeurer** [comment]
Hm, the tests fail only for python2. Can we check that later?

**Comment by amueller** [comment]
yeah. merge this now into your PR?

**Comment by amueller** [comment]
I can't reproduce the issue on python2.7 :(

**Comment by mfeurer** [comment]
Okay, I'll merge this now. I'll opened #184 about dropping python2.7 support.
